quic/s2n-quic-core/src/path/mtu.rs

```diff
@@ -232,11 +231,102 @@ impl Display for MtuError {
 #[cfg(feature = "std")]
 impl std::error::Error for MtuError {}
 
+/// Information about the path that may be used when generating or
+/// validating MTU configuration.
+#[non_exhaustive]
+pub struct PathInfo<'a> {
+    pub remote_address: event::api::SocketAddress<'a>,
+}
+
+impl<'a> PathInfo<'a> {
+    #[inline]
+    #[doc(hidden)]
+    pub fn new(remote_address: &'a inet::SocketAddress) -> Self {
+        PathInfo {
+            remote_address: remote_address.into_event(),
+        }
+    }
+}
+
+/// MTU configuration manager.
+#[derive(Copy, Clone, Debug)]
+pub struct Manager<E: mtu::Endpoint> {
+    provider: E,
+    endpoint_mtu_config: Config,
+}
+
+impl<E: mtu::Endpoint> Manager<E> {
+    pub fn new(provider: E) -> Self {
+        Manager {
+            provider,
+            // Instantiate the Manager with default values since the endpoint is
+            // created before the IO provider (IO provider sets the actual config `set_mtu_config()`).
+            endpoint_mtu_config: Default::default(),
+        }
+    }
+
+    pub fn config(&mut self, info: &PathInfo) -> Result<Config, MtuError> {
+        if let Some(conn_config) = self.provider.on_path(info, self.endpoint_mtu_config) {
+            ensure!(conn_config.is_valid(), Err(MtuError));
+            ensure!(
+                u16::from(conn_config.max_mtu) <= u16::from(self.endpoint_mtu_config.max_mtu()),
+                Err(MtuError)
+            );
+
+            Ok(conn_config)
+        } else {
+            Ok(self.endpoint_mtu_config)
+        }
+    }
+
+    pub fn set_endpoint_config(&mut self, config: Config) {
+        self.endpoint_mtu_config = config;
+    }
+
+    pub fn endpoint_config(&self) -> &Config {
+        &self.endpoint_mtu_config
+    }
+}
+
+/// Creates MTU config for the given path.
+pub trait Endpoint: 'static + Send {
+    /// Provide path specific MTU config.
+    ///
+    /// Application must ensure that `max_mtu <= info.mtu_config.max_mtu`.
```

Comment:
Can we put an explicit note that this can be called multiple times per connection

Reply:
Do you mean multiple times per endpoint (which is prob more common)? It can technically also be called multiple times per conn during connection migration.

Reply:
Actually its called for each path.. Ill adjust the wording.